gematria/datasets/exegesis_benchmark_lib.cc

```diff
@@ -0,0 +1,371 @@
+// Copyright 2024 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gematria/datasets/exegesis_benchmark_lib.h"
+
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <optional>
+#include <string_view>
+#include <utility>
+#include <vector>
+
+#include "gematria/datasets/find_accessed_addrs.h"
+#include "gematria/llvm/disassembler.h"
+#include "gematria/utils/string.h"
+#include "llvm/ADT/APInt.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/IR/InlineAsm.h"
+#include "llvm/MC/MCContext.h"
+#include "llvm/MC/MCInst.h"
+#include "llvm/MC/MCInstPrinter.h"
+#include "llvm/MC/TargetRegistry.h"
+#include "llvm/Support/Errc.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/JSON.h"
+#include "llvm/tools/llvm-exegesis/lib/BenchmarkCode.h"
+#include "llvm/tools/llvm-exegesis/lib/BenchmarkResult.h"
+#include "llvm/tools/llvm-exegesis/lib/BenchmarkRunner.h"
+#include "llvm/tools/llvm-exegesis/lib/LlvmState.h"
+#include "llvm/tools/llvm-exegesis/lib/RegisterValue.h"
+#include "llvm/tools/llvm-exegesis/lib/ResultAggregator.h"
+#include "llvm/tools/llvm-exegesis/lib/SnippetRepetitor.h"
+#include "llvm/tools/llvm-exegesis/lib/Target.h"
+
+using namespace llvm;
+using namespace llvm::exegesis;
+
+namespace gematria {
+
+ExegesisBenchmark::ExegesisBenchmark(LLVMState &&State)
+    : ExegesisState(std::move(State)) {
```

Comment:
**tl;dr:** The `std::move()` is not needed.

**Longer:** The only thing that `std::move()` does is that it converts a reference to an rvalue reference. When `State` is already an rvalue-reference, this is a no-op.

Reply:
Will address this separately given the patch for that already landed.

Reply:
Actually, when trying to fix that, `clangd` is giving me an error about a deleted copy constructor. 

Not entirely sure why given the `std::move` is necessary here given we're basically `static_cast<>`ing a r-value reference to a r-value reference (like you mentioned).

Reply:
Ah, right.

From https://en.cppreference.com/w/cpp/language/value_category#lvalue:
> Even if the variable's type is rvalue reference, the expression consisting of its name is an lvalue expression

Reply:
Today I learned. C++ being complicated again. Thanks for pulling that up!